experimental/cuda2/memory_pools.c

```diff
@@ -276,17 +276,27 @@ iree_status_t iree_hal_cuda2_memory_pools_dealloca(
   IREE_TRACE_ZONE_APPEND_VALUE(
       z0, (int64_t)iree_hal_buffer_allocation_size(buffer));
 
-  // Try to schedule the buffer for freeing.
-  CUdeviceptr device_ptr = iree_hal_cuda2_buffer_device_pointer(buffer);
-  iree_status_t status = IREE_CURESULT_TO_STATUS(
-      pools->cuda_symbols, cuMemFreeAsync(device_ptr, stream),
-      "cuMemFreeAsync");
+  // Only process the request if the buffer came from an async pool.
+  // We may get requests for deallocations on ones that didn't if one part of
+  // the application allocated the buffer synchronously and another deallocated
+  // it asynchronously.
+  iree_status_t status = iree_ok_status();
+  if (iree_hal_cuda2_buffer_type(buffer) == IREE_HAL_CUDA_BUFFER_TYPE_ASYNC) {
+    // Try to schedule the buffer for freeing.
+    CUdeviceptr device_ptr = iree_hal_cuda2_buffer_device_pointer(buffer);
+    status = IREE_CURESULT_TO_STATUS(pools->cuda_symbols,
+                                     cuMemFreeAsync(device_ptr, stream),
+                                     "cuMemFreeAsync");
 
-  // Drop the release callback so that we don't try to double-free the buffer.
-  iree_hal_cuda2_buffer_drop_release_callback(buffer);
+    // Drop the release callback so that we don't try to double-free the buffer.
+    iree_hal_cuda2_buffer_drop_release_callback(buffer);
```

Comment:
if the cuMemFreeAsync fails we want to still free the buffer synchronously... which this doesn't do, so good catch I'll fix that :)
it's only valid to dealloca a buffer once so there can't be any other deallocas happening and the only thing that relies on the callback is the buffer destroy when all references drop to zero but the caller still owns a reference so the earliest it could happen would be after the function returns on the calling thread.